drivers/lock_queue.py

```diff
@@ -0,0 +1,121 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# Miscellaneous utility functions
+#
+import os
+import pickle
+import sys
+import time
+
+import lock
+import util
+
+
+DEBUG_LOG = True
+
+def debug_log(msg):
+    if DEBUG_LOG:
+        util.SMlog("LockQueue: " + msg)
+
+def get_process_start_time(pid):
+    proc_file = f"/proc/{pid}/stat"
+    with open(proc_file, 'r') as f:
+        return f.read().split(')')[-1].split(' ')[20]
+
+def process_is_valid(pid, start_time):
+    proc_file = f"/proc/{pid}/stat"
+
+    try:
+        if start_time != get_process_start_time(pid):
+            debug_log(f"Process {pid} has incorrect start time real:{get_process_start_time(pid)} vs expected:{start_time}")
+            return False
+    except FileNotFoundError:
+        debug_log(f"Process {pid} is dead")
+        return False
+
+    return True
+
+class LockQueue:
+    def __init__(self, name):
+        self.name = name
+        self._queue_lock = lock.Lock(name, f"ql-{name}")
+        self._action_lock = lock.Lock(name, f"al-{name}")
+
+        # Open file to hold the process queue
+        self._mem = f"/tmp/mem-{name}"
+        if not os.path.exists(self._mem):
+            # If the queue did not exist then initialize with empty queue
+            self.save_queue([])
+
+    def load_queue(self):
+        with open(self._mem, "rb") as f:
+            queue = pickle.load(f)
+        debug_log("load_queue {}".format(queue))
+        return queue
+
+    def save_queue(self, queue):
+        with open(self._mem, "w+b") as f:
+            pickle.dump(queue, f)
+        debug_log("save_queue {}".format(queue))
+
+    def push_into_process_queue(self):
+        self._queue_lock.acquire()
+
+        queue = self.load_queue()
+        queue.append((os.getpid(), get_process_start_time(os.getpid())))
+        self.save_queue(queue)
+
+        self._queue_lock.release()
+
+    def __enter__(self):
+        # Add ourselves to the process queue.
+        self.push_into_process_queue()
+
+        # Keep reading the process queue until we are at the front
+        while True:
+            self._queue_lock.acquire()
+            queue = self.load_queue()
+            front_pid, front_start_time = queue.pop(0)
+            print(f"Testing for PID {front_pid}")
+            if front_pid == os.getpid():
+                # We are at the front, it is now our turn to wait on the action lock
+                # and then do our work
+                debug_log(f"{front_pid} taking action lock")
+                self._action_lock.acquire()
```

Comment:
If another process already holds the action lock, then the current process will have to wait until it can release the queue lock. In which case, any other processes wanting to take out locks won't even get onto the queue until that happens, and the order in which they end up on the queue will depend on flock, which I thought were didn't trust on that score. I don't know how this would pan out in practice, but the locking might not be as fair as we hope.

Reply:
Consider what happens when the process holding the action lock is killed. As little as possible should be required of other processes when that happens. In this case, nothing is required as there is nothing to clean up, and no handlers need to be installed to clean up after non-KILL signals. Cleanup only needs to worry about waiters who died while waiting, which should be much less common.

Also, your solution would require regular checking of the queue head while the action lock is held, which basically means you've got a load of processes constantly checking after short sleeps instead of properly asleep in the kernel. You've implemented a spinlock. Let's not do that.

I am well aware that this mechanism isn't perfectly fair. I proposed a method which is, and only requires one lock, but it's not what got implemented (arguably its check is too heavyweight). So it goes. In the case of a lot of queued waiters, processes already in the queue will sleep between checks where new arrivals won't. so that should help.